juju/model.py

```diff
@@ -151,6 +171,26 @@ def _load(self):
         # in-memory cache as well as calling out to relation-set.
         return self._backend.relation_get(self.relation_id, self.unit.name)
 
+    def __setitem__(self, key, value):
+        if not self._is_mutable:
+            raise TypeError(f'Cannot set relation data for {self.unit.name}')
```

Comment:
I think we discussed this convention in the week before you joined: I suggest we stick to lowercased error messages, since these usually end up being concatenated together in strings like:
```
error: something bad happened: the fooer wasn't found: connection lost
```
This looks much nicer if it's all consistently lowercased than if every first word is upper cased.

Then, this doesn't look like a type error. It's relation data, and some relations do support mutation, so it's more of a business logic misunderstanding than a programming error with a bad type. I suggest creating a more specific RelationDataError and using it for these two cases, so we can also test these cases more specifically.